src/gap_to_julia.jl

```diff
@@ -78,164 +94,153 @@ function gap_to_julia(::Type{Rational{T}}, x::Int64) where T <: Integer
 
 ## Floats
-function gap_to_julia( ::Type{Float64}, obj::GapObj)
-    if ! Globals.IsFloat(obj)
-        throw(ArgumentError("<obj> is not a MacFloat"))
-    end
-    return ValueMacFloat(obj)
+function gap_to_julia(::Type{Float64}, obj::GapObj)
+    Globals.IsIEEE754FloatRep(obj) && return ValueMacFloat(obj)::Float64
+    throw(ConversionError(obj, Float64))
 end
 
-gap_to_julia( ::Type{Float32}, obj::GapObj)  = Float32(gap_to_julia(Float64,obj))
-gap_to_julia( ::Type{Float16}, obj::GapObj)  = Float16(gap_to_julia(Float64,obj))
-gap_to_julia( ::Type{BigFloat}, obj::GapObj) = BigFloat(gap_to_julia(Float64,obj))
+gap_to_julia(::Type{T}, obj::GapObj) where T<:AbstractFloat = T(gap_to_julia(Float64, obj))
 
 ## Chars
 function gap_to_julia( ::Type{Char}, obj::GapObj)
-    if ! Globals.IsChar( obj )
-        throw(ArgumentError("argument is not a character object"))
-    end
-    return Char( Globals.INT_CHAR(obj ) )
+    Globals.IsChar(obj) && return Char(Globals.INT_CHAR(obj))
+    throw(ConversionError(obj, Char))
 end
 
 function gap_to_julia( ::Type{Cuchar}, obj::GapObj)
-    if ! Globals.IsChar( obj )
-        throw(ArgumentError("argument is not a character object"))
-    end
-    return trunc( Cuchar, Globals.INT_CHAR(obj ) )
+    Globals.IsChar(obj) && return trunc(Cuchar, Globals.INT_CHAR(obj))
+    throw(ConversionError(obj, Cuchar))
 end
 
-## Strings and symbols
-function gap_to_julia(::Type{String},obj::GapObj)
-    if Globals.IsStringRep(obj)
-        return CSTR_STRING(obj)
-    elseif Globals.IsString(obj)
-        return CSTR_STRING( Globals.CopyToStringRep( obj ) )
-    else
-        throw(ArgumentError("<obj> is not a string"))
-    end
+## Strings
+function gap_to_julia(::Type{String}, obj::GapObj)
+    Globals.IsStringRep(obj) && return CSTR_STRING(obj)
+    Globals.IsString(obj) && return CSTR_STRING(Globals.CopyToStringRep(obj))
+    throw(ConversionError(obj, String))
 end
-gap_to_julia(::Type{AbstractString},obj::GapObj) = gap_to_julia(String,obj)
-gap_to_julia(::Type{Symbol},obj::GapObj) = Symbol(gap_to_julia(String,obj))
+gap_to_julia(::Type{T},obj::GapObj) where T <: AbstractString = convert(T, gap_to_julia(String,obj))
 
-gap_to_julia(type_obj,obj,recursion_dict) = gap_to_julia(type_obj,obj)
+## Symbols
+gap_to_julia(::Type{Symbol},obj::GapObj) = Symbol(gap_to_julia(String,obj))
 
-## Convert GAP string to Array{UInt8,1}
-function gap_to_julia( ::Type{Array{UInt8,1}}, obj :: GapObj )
-    ## convert strings to uint8 lists, if requested
-    if Globals.IsStringRep( obj )
-        array = UNSAFE_CSTR_STRING( obj )
-        return deepcopy(array)
-    elseif Globals.IsList( obj )
-        return Array{UInt8,1}(map(i->gap_to_julia(UInt8,obj[i]),1:length(obj)))
-    else
-        throw(ArgumentError("<obj> is not a list"))
-    end
+## Convert GAP string to Vector{UInt8} (==Array{UInt8,1})
+function gap_to_julia(::Type{Vector{UInt8}}, obj::GapObj)
+    Globals.IsStringRep(obj) && return deepcopy(UNSAFE_CSTR_STRING(obj))
```

Comment:
as I said above: I'd move `deepcopy` into `UNSAFE_CSTR_STRING` and then rename `UNSAFE_CSTR_STRING`